File: 0020_Valid_Parenthesis/0020_Valid_Parenthesis.ts
function isValid(s: string): boolean {
    var stack: string[] = [];
    const opp = new Map([
        [')', '('],
        ['}', '{'],
        [']', '[']
    ])

    for (let ch of s) {
        // Found Closing Parenthesis
        if (opp.has(ch)) {
            // Closing comes earlier than opening -> Invalid
            if (stack.length === 0) {
                return false;
            }
            
            const lastSeen = stack.pop();

            // Unmatching parenthesis -> Invalid
            if (lastSeen !== opp.get(ch)) {
                return false;
            }
        }

        // Opening Parenthesis
        else {
            stack.push(ch);
        }
    }

    return stack.length === 0;
};

File: 0020_Valid_Parenthesis/0020_Valid_Parenthesis.py
class Solution:
    # Time O(n)
    # Space O(n)
    def isValid(self, s: str) -> bool:
        stack = []
        opp = {')' : '(', '}': '{', ']' : '['}
        
        # Odd Length -> Cannot be valid parenthesis
        if len(s) % 2 == 1:
            return False

        for elem in s:
            # Found closing
            if elem in opp:
                # Closing parenthesis was found ealier than opening
                if len(stack) == 0:
                    return False

                # get last seen opening from stack
                lastSeen = stack.pop()
                
                # Found Unmatching parenthesis
                if lastSeen != opp[elem]:
                    return False

            # Save opening parenthesis into stack
            else:
                stack.append(elem)
            
        # Everything matched (stack empty) -> Valid Parenthesis
        return len(stack) == 0